tasks/cephfs/test_strays.py

```diff
@@ -0,0 +1,465 @@
+        self.assertEqual(self.get_mdc_stat("strays_created"), 1)
+
+        # Wait, see that data objects are still present (i.e. that the
+        # stray did not advance to purging given time)
+        time.sleep(30)
+        self.assertTrue(self.fs.data_objects_present(ino, size_mb * 1024 * 1024))
+        self.assertEqual(self.get_mdc_stat("strays_purged"), 0)
+
+        # See that before reintegration, the inode's backtrace points to a stray dir
+        self.fs.mds_asok(['flush', 'journal'])
+        self.assertTrue(self.get_backtrace_path(ino).startswith("stray"))
+
+        # Do a metadata operation on the remaining link (mv is heavy handed, but
+        # others like touch may be satisfied from caps without poking MDS)
+        self.mount_a.run_shell(["mv", "file_b", "file_c"])
+
+        # See the reintegration counter increment
+        # This should happen as a result of the eval_remote call on
+        # responding to a client request.
+        self.wait_until_equal(
+            lambda: self.get_mdc_stat("strays_reintegrated"),
+            expect_val=1, timeout=60, reject_fn=lambda x: x > 1
+        )
+
+        # Flush the journal
+        self.fs.mds_asok(['flush', 'journal'])
+
+        # See that the backtrace for the file points to the remaining link's path
+        post_reint_bt = self.fs.read_backtrace(ino)
+        self.assertEqual(post_reint_bt['ancestors'][0]['dname'], "file_c")
+
+        # See that the number of strays in existence is zero
+        self.assertEqual(self.get_mdc_stat("num_strays"), 0)
+
+        # Now really delete it
+        self.mount_a.run_shell(["rm", "-f", "file_c"])
+        self.wait_until_equal(
+            lambda: self.get_mdc_stat("strays_purged"),
+            expect_val=1, timeout=60, reject_fn=lambda x: x > 1
+        )
+        self.assert_purge_idle()
+        self.assertTrue(self.fs.data_objects_absent(ino, size_mb * 1024 * 1024))
+
+        # We caused the inode to go stray twice
+        self.assertEqual(self.get_mdc_stat("strays_created"), 2)
+        # One time we reintegrated it
+        self.assertEqual(self.get_mdc_stat("strays_reintegrated"), 1)
+        # Then the second time we purged it
+        self.assertEqual(self.get_mdc_stat("strays_purged"), 1)
+
+    def test_mv_hardlink_cleanup(self):
+        """
+        That when doing a rename from A to B, and B has hardlinks,
+        then we make a stray for B which is then reintegrated
+        into one of his hardlinks.
+        """
+        # Create file_a, file_b, and a hardlink to file_b
+
+        # mv file_a file_b
+
+        # Proceed as with test_hardlink_reintegration
+
+    def test_migration_on_shutdown(self):
+        """
+        That when an MDS rank is shut down, any not-yet-purging strays
+        are migrated to another MDS's stray dir.
+        """
+
+        # XXX interestingly, export dir on an empty dir doesn't work.
+        # perhaps because it has no actual dirfrag at that stage?
```

Comment:
At the time it came as a surprise, now I don't think it's an issue as it just reflects the fact that no dirfrag necessarily exists until some dentry is created.  Will rephrase